keps/sig-node/2238-liveness-probe-grace-period/README.md

```diff
@@ -0,0 +1,544 @@
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How does this feature react if the API server and/or etcd is unavailable?**
+
+* **What are other known failure modes?**
+  For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+
+* **What steps should be taken if SLOs are not being met to determine the problem?**
+
+[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
+[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
+
+## Implementation History
+
+- 2021-01-07: Initial draft KEP
+
+## Drawbacks
+
+The main drawback of this proposal is that it requires a (compatible) API
+change.
+
+## Alternatives
+
+From [#64715](https://github.com/kubernetes/kubernetes/issues/64715#issuecomment-756173979):
+
+Infer the maximum grace period for liveness from `num failures * probe period`,
+implying that a user with short liveness reaction wants a fast reaction from
+their pod (@smarterclayton)
+
+- PRO: Any user unaware of the coupling reacts to liveness failures faster,
+  normal pods (including reason 2 above) are less likely to be impacted
+- CON: If we later want to change the API to indicate this field, the default
+  behavior makes less sense (i.e. complicates choosing the option above)
+- CON: A user with a long grace period who wants liveness to take more than
+  tens of seconds (not sure why they would) suddenly sees that behavior
+  impacted
+- CON: This sort of implicit calculation feels pretty arbitrary (we generally
+  make defaults explicit vs calculated)
+
+Introduce a new toleration, similar to execute toleration for node not ready
+(how long the pod should be left on the node while unready before being
+evicted), that controls how long liveness tolerates (@smarterclayton)
+
+- PRO: avoids API change
+- CON: basically a hack to avoid an API change, tolerations for execution don't
+  really make sense for conditional stuff
+
+Since liveness probe failures indicate that a container is unhealthy and should
+be terminated, update the behaviour of liveness probes to terminate the pod
+immediately, or with some default grace period (@sjenning)
+
+- PRO: avoids API change
```

Comment:
I will add another CON: Kubernetes changes behavior in a surprising way, potentially breaking existing apps that depend on grace period for crash safety.  That CON is sufficient for me to NACK such a change, because the cost of breaking existing users in these scenarios is potentially very high and would happen in the worst scenario (terminationGracePeriod is intended to allow applications a bounded time window to shut down).

I just want to record that here so that in the future we treat crash safety under SIGTERM as it was intended all the way back in https://github.com/kubernetes/kubernetes/issues/2789 (https://github.com/kubernetes/kubernetes/pull/8641 is where we actually say "don't SIGKILL my database" in an issue, it was always implicit).